feat(dataworker): Add function to load and format data used in all functions

nicholaspai opened the pull request:
All data worker functions will need some combination of:
- unfilled deposits grouped by destination chain
    - Will be used to create `SlowRelayRoots` and `PoolRebalanceRoots`
- validated fill data grouped by repayment chain
    - Will be used to create `RelayerRefundRoots` and `PoolRebalanceRoots`

nicholaspai commented on src/clients/SpokePoolClient.ts:
```diff
@@ -30,7 +31,11 @@ export class SpokePoolClient {
+    return Object.values(this.fills).flat();
+  }
+
+  getFillsForDestinationChain(destinationChainId: number): Fill[] {
```
I added `getFills` because I think I just need ALL the fill data for a given destination. Why would filtering fills on destination chain (i.e. calling `getFillsForDestinationChain`) ever be useful? When would fills for a given spoke pool ever have a different destination chain than the spoke pool's chain?

chrismaree replied:
`getFillsForDestinationChain` is currently not used but you do raise a good point...the data structure:
```ts
private fills: { [DestinationChainId: number]: Fill[] } = {};
```
does not actually make sence as all fills seen on this client will, be definition, be at the destination chain of _this_ client.

`getDepositsForDestinationChain` is obviously diffrent as it's saying: from the perspective of _this_ client's spokePool the target is this other chain.

nicholaspai replied:
Yeah so should we remove the `[DestinationChainId]` key?

nicholaspai replied:
We could then flatten the fills to be an array of `Fill[]`

nicholaspai replied:
yeah so far I haven't had to use `getFillsForRepaymentChain` so I won't add that in this PR. But I will flatten the `fills` structure a bit.

nicholaspai commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,113 @@ export class Dataworker {
+            for (const deposit of depositsForDestinationChain) {
+              // @dev: It doesn't matter which client we call validateFillForDeposit() on as the logic is 
+              // chain agnostic.
+              if (destinationClient.validateFillForDeposit(fill, deposit)) return true;
```
IIUC, `validateFillForDeposit` is chain agnostic. Should we move this to a more common client, perhaps to a `ClientTools` class?

chrismaree commented on src/clients/RateModelClient.ts:
```diff
@@ -54,11 +54,6 @@ export class RateModelClient {
     return this.rateModelDictionary.getRateModelForBlockNumber(l1Token, blockNumber);
   }
 
-  async validateRealizedLpFeePctForFill(fill: Fill, deposit: Deposit) {
```
👍

chrismaree commented on src/clients/SpokePoolClient.ts:
```diff
@@ -56,7 +61,7 @@ export class SpokePoolClient {
   }
 
-  getValidUnfilledAmountForDeposit(deposit: Deposit) {
+  getValidUnfilledAmountForDeposit(deposit: Deposit): BigNumber {
```
thanks for adding the return type. useful. we will have a separate PR later to go in and make this compliant all over the place.

The PR I'm working on also adds a bunch of additional logging spots.

nicholaspai replied:
Sounds good I just added where I used it

chrismaree commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+
+    for (const originChainId of Object.keys(this.spokePoolClients)) {
+      const originClient = this.spokePoolClients[originChainId];
+      if (!originClient.isUpdated())
+        throw new Error(`origin spokepoolclient with chain ID ${originChainId} not updated`);
```
silly nit but let's try keep the total line length on these kinds of errors down. ideally it's just a one liner.

```suggestion
      if (!originClient.isUpdated()) throw new Error(`origin spokepoolclient on chain ${originChainId} not updated`);
```

chrismaree commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+        if (originChainId === destinationChainId) continue;
+
+        const destinationClient = this.spokePoolClients[destinationChainId];
+        if (!destinationClient.isUpdated())
```
same nit about this log. one line pllleeease

chrismaree commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+
+        const unfilledDepositsForDestinationChain = depositsForDestinationChain
+          .map((deposit) => {
+            return { data: deposit, unfilledAmount: destinationClient.getValidUnfilledAmountForDeposit(deposit) };
```
why key this by `data`? would it not be better to key it by just `deposit` then it will become:
```ts
return { deposit, unfilledAmount: destinationClient.getValidUnfilledAmountForDeposit(deposit) };
```

nicholaspai replied:
sure

chrismaree commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+      }
+    }
+
+    return {
+      fillsToRefund,
+      unfilledDeposits,
+    };
```
```suggestion
    return { fillsToRefund, unfilledDeposits };
```

chrismaree commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+          destinationChainId,
+        });
+
+        for (const fill of validFillsOnDestinationChain) {
+          assign(fillsToRefund, [fill.repaymentChainId, fill.relayer], [fill]);
+        }
```
might need to be listed. not sure if this is too long.
```suggestion
        for (const fill of validFillsOnDestinationChain)assign(fillsToRefund, [fill.repaymentChainId, fill.relayer], [fill]);
```

nicholaspai replied:
would a `forEach` make you happy?

chrismaree replied:
what you have is fine. just you dont need the extra pair of `{` & `}` :)

nicholaspai replied:
gotcha i ended up using `forEach`

chrismaree commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+        const validFillsOnDestinationChain = fillsOnDestinationChain.filter((fill) => {
+          // For each fill, see if we can find a deposit sent from the origin client that matches it.
+          for (const deposit of depositsForDestinationChain) {
+            // @dev: It doesn't matter which client we call validateFillForDeposit() on as the logic is
+            // chain agnostic.
+            // @dev: All of the deposits returned by `getDepositsForDestinationChain` will include the expected realized
+            // lp fee % for the deposit quote time. If this fill does not have the same realized lp fee %, then it will
+            // be ignored.
```
this commenting is non-standard with the `@dev` flag. we should not mix and match commenting styles methinks.
```suggestion
            // Note: it doesn't matter which client we call validateFillForDeposit() on as the logic is chain agnostic.
            // All of the deposits returned by getDepositsForDestinationChain will include the expected realized
            // lp fee % for the deposit quote time. If this fill does not have the same realized lp fee %, then it will be ignored.
```

chrismaree commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+import { deploySpokePoolWithToken, enableRoutesOnHubPool, destinationChainId, originChainId, sinon } from "./utils";
```
some of these should be moved to the `const` import such as `destinationChainId, originChainId` then you'll be able to pull the imports from the following line into this line and decrease the line count in this bulky import.

nicholaspai replied:
Fixed

chrismaree commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+
+  it("Default conditions", async function () {
+    // Throws error if spoke pool client not updated.
+    expect(() => dataworkerInstance._loadData()).to.throw(/not updated/);
```
ooo this is great!

chrismaree commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+  it("Returns unfilled deposits", async function () {
+    await updateAllClients();
+
+    const deposit1 = await buildDeposit(
+      spokePool_1,
+      erc20_1,
+      depositor,
+      depositor,
+      destinationChainId,
+      amountToDeposit
+    );
```
I know you are going to hate me for this but do you think we can decrease the line length on this so it is not 8 lines repeated over and over again? the simplest modification I can see is to change `buildDeposit` to make the `depositor` and `recipient` the same address. this is basically what I did in `fillWithRealizedLpFeePct` here: https://github.com/across-protocol/relayer-v2/blob/b39b050a7d6b93fae3417aebf9237c0a4962b3a2/test/Relayer.UnfilledDeposits.ts#L143

nicholaspai replied:
cool done where possible

chrismaree commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+      { ...deposit2, realizedLpFeePct: deposit2.realizedLpFeePct.div(toBN(2)) },
+      0.25
+    );
+    // @dev: This fill has identical deposit data to fill2 except for the destination token being different
```
```suggestion
    // This fill has identical deposit data to fill2 except for the destination token being different
```

nicholaspai replied:
Fixed

chrismaree commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+    });
+
+    // Submit fills without matching deposits. These should be ignored by the client.
+    // @dev: Switch the deposit data to make fills invalid.
```
same nit as before around using the `@dev` syntax.
```suggestion
    // Switch the deposit data to make fills invalid.
```

nicholaspai replied:
Fixed

chrismaree commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+    // Submit fills that match deposit in all properties except for realized lp fee % or l1 token. These should be
+    // ignored because the rate model client deems them invalid. These are the two properties added to the deposit
+    // object by the spoke pool client.
+    // @dev: This fill has identical deposit data to fill2 except for the realized lp fee %.
```
```suggestion
    // This fill has identical deposit data to fill2 except for the RealizedLpFeePct.
```

nicholaspai replied:
Fixed

chrismaree commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+    _destinationChainId,
+    _amountToDeposit
+  );
+  return {
```
is the whole point of this method just to append extra data onto the deposit object? if so, this would be useful in other places I do something similar here: https://github.com/across-protocol/relayer-v2/blob/b39b050a7d6b93fae3417aebf9237c0a4962b3a2/test/Relayer.UnfilledDeposits.ts#L70 where I append these other props directly onto the object in the exact same format you are doing it here. if you refactored this into utils we could simplify those other tests as well.

nicholaspai replied:
Also applied this util function in all other tests

chrismaree commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+          })
+          .filter((deposit) => deposit.unfilledAmount.gt(0));
+
+        if (unfilledDepositsForDestinationChain.length > 0) {
+          assign(unfilledDeposits, [destinationChainId], unfilledDepositsForDestinationChain);
+        } else {
+          this.logger.debug({
+            at: "Dataworker",
+            message: `All deposits are filled`,
+            originChainId,
+            destinationChainId,
+          });
+        }
```
```suggestion
        if (unfilledDepositsForDestinationChain.length > 0) assign(unfilledDeposits, [destinationChainId], unfilledDepositsForDestinationChain);
        else
          this.logger.debug({
            at: "Dataworker",
            message: `All deposits are filled`,
            originChainId,
            destinationChainId,
          });
```

chrismaree reviewed (commented):
this is really awesome work @nicholaspai ! I'm sorry I left so many nits...I just really want to keep this repo as consistant and clean as posible as we add more and more complex code. Please try some of the refactors I proposed and I'll re review. I want to go through the logic on the filtering a bit more closely once the other changes are in.

nicholaspai commented on src/dataworker/Dataworker.ts:
```diff
@@ -1,9 +1,17 @@
-// @notice Constructs roots to submit to HubPool on L1
+// @notice Constructs roots to submit to HubPool on L1. Fetches all data synchronously from SpokePool/HubPool clients
+// so this class assumes that those upstream clients are already updated and have fetched on-chain data from RPC's.
+type UnfilledDeposit = {
```
Ah good catch

nicholaspai replied:
I don't really understand the diff between types and interfaces but [this post](https://stackoverflow.com/questions/37233735/interfaces-vs-types-in-typescript) seems to suggest that types are much better for what we're doing since we're not trying to extend these interfaces in our class implementations.

nicholaspai replied:
according to that style these should be interfaces then so I'll change them and move to `interfaces/SpokePool.ts` for now

nicholaspai commented on src/dataworker/Dataworker.ts:
```diff
@@ -13,40 +21,134 @@ export class Dataworker {
+    const unfilledDeposits: UnfilledDeposits = {};
+    const fillsToRefund: FillsToRefund = {};
+
+    for (const originChainId of Object.keys(this.spokePoolClients)) {
```
Good call. Done

nicholaspai commented on test/Dataworker.loadData.ts:
```diff
@@ -0,0 +1,301 @@
+  _amountToDeposit: BigNumber = amountToDeposit
+): Promise<Deposit> {
+  const _deposit = await deposit(
+    spokePool,
+    tokenToDeposit,
+    recipient,
+    depositor,
+    _destinationChainId,
+    _amountToDeposit
+  );
+  return {
+    ..._deposit,
+    destinationToken: hubPoolClient.getDestinationTokenForDeposit(_deposit),
+    realizedLpFeePct: await rateModelClient.computeRealizedLpFeePct(_deposit, l1Token.address),
+  };
+}
+
+async function buildFill(
+  spokePool: Contract,
+  destinationToken: Contract,
+  recipient: SignerWithAddress,
+  depositor: SignerWithAddress,
+  relayer: SignerWithAddress,
+  deposit: Deposit,
+  pctOfDepositToFill: number
+): Promise<Fill> {
+  return await fillRelay(
+    spokePool,
+    destinationToken,
+    recipient,
+    depositor,
+    relayer,
+    deposit.depositId,
+    deposit.originChainId,
+    deposit.amount,
+    deposit.amount.mul(toBNWei(pctOfDepositToFill)).div(toBNWei(1)),
+    deposit.realizedLpFeePct,
+    deposit.relayerFeePct
+  );
+}
```
Hmmm probably yeah

nicholaspai commented:
> this is really awesome work @nicholaspai ! I'm sorry I left so many nits...I just really want to keep this repo as consistant and clean as posible as we add more and more complex code. Please try some of the refactors I proposed and I'll re review. I want to go through the logic on the filtering a bit more closely once the other changes are in.

Added all syntactical nits and added the helper functions to `test/utils`! Hopefully now you can review the logic but please do point out other stylistical inconsistencies

chrismaree reviewed (approved):
This is sick! thanks for going through all my feedback.
